src/debug-generator/debug-generator.c

```diff
@@ -20,20 +22,95 @@
 static char *arg_default_unit = NULL;
 static char **arg_mask = NULL;
 static char **arg_wants = NULL;
 static bool arg_debug_shell = false;
 static char *arg_debug_tty = NULL;
 static char *arg_default_debug_tty = NULL;
+static uint32_t arg_breakpoints = 0;
 
 STATIC_DESTRUCTOR_REGISTER(arg_default_unit, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_mask, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_wants, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_debug_tty, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_default_debug_tty, freep);
 
+static const struct BreakpointInfo breakpoint_info_table[_BREAKPOINT_TYPE_MAX] = {
+        { BREAKPOINT_PRE_UDEV,          "pre-udev",        "breakpoint-pre-udev.service",        true,  true  },
+        { BREAKPOINT_PRE_SYSROOT_MOUNT, "pre-mount",       "breakpoint-pre-mount.service",       true,  false },
+        { BREAKPOINT_PRE_SWITCH_ROOT,   "pre-switch-root", "breakpoint-pre-switch-root.service", true,  false },
+};
+
+static BreakpointType parse_breakpoint_from_string_one(const char *s) {
+        assert(s);
+
+        for (size_t i = 0; i < ELEMENTSOF(breakpoint_info_table); i++)
+                if (streq(breakpoint_info_table[i].name, s))
+                        return breakpoint_info_table[i].type;
+
+        return _BREAKPOINT_TYPE_INVALID;
+}
+
+static int parse_breakpoint_from_string(const char *s, uint32_t *ret_breakpoints) {
+        uint32_t breakpoints = 0;
+
+        assert(ret_breakpoints);
+
+        /* Empty value? set default breakpoint */
+        if (!s) {
+                if (in_initrd())
+                        breakpoints |= (1 << BREAKPOINT_PRE_SWITCH_ROOT);
+                else
+                        log_warning("No default breakpoint defined in the main system, ignoring.");
+        } else {
+                int r;
+
+                for (;;) {
+                        _cleanup_free_ char *t = NULL;
+                        BreakpointType tt;
+
+                        r = extract_first_word(&s, &t, ",", EXTRACT_DONT_COALESCE_SEPARATORS);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        tt = parse_breakpoint_from_string_one(t);
+                        if (tt < 0)
+                                return log_warning_errno(SYNTHETIC_ERRNO(EBADRQC), "Invalid breakpoint value '%s'", t);
+
+                        if (in_initrd() && !breakpoint_info_table[tt].valid_in_initrd)
+                                log_warning("Breakpoint '%s' not valid in the initrd, ignoring.", t);
+                        else if (!in_initrd() && !breakpoint_info_table[tt].valid_in_main_system)
+                                log_warning("Breakpoint '%s' not valid in the main system, ignoring.", t);
+                        else
+                                breakpoints |= (1 << tt);
+                }
+        }
+
+        *ret_breakpoints |= breakpoints;
```

Comment:
we have this rules that output params should be named ret_xyz, but this is not an output param you are currently using it: it's also an input param, given that you OR into it, instead of overriding it. 

in order to minimize surprises I think you should assign it, it thus remove the `|`.

Reply:
i mean, in general I think ORing the requests in case systemd.break= is used multiple times, but I think that should be done outside of parse_breakpoint_from_string() and instead in parse_proc_cmdline_item() below